### ISSUE TITLE
Warn on access denied errors when deleting from lib-bkp during uninstall

### ISSUE THREAD

**Issue by vorou** [chocolatey/choco#263]
It happens after my `chocolateyUninstall.ps1` is finished. I've tried to log handles, didn't find the file there. I can remove it manually after uninstalling is finished.

```
Attempting to delete directory "C:\ProgramData\chocolatey\lib-bkp\rbtray".
This is try 1/3. Retrying after 300 milliseconds.
 Error converted to warning:
 System.UnauthorizedAccessException: Access to the path 'RBHook.dll' is denied.
   at System.IO.Directory.DeleteHelper(String fullPath, String userPath, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.Directory.Delete(String fullPath, String userPath, Boolean recursive, Boolean checkHost)
   at chocolatey.infrastructure.tolerance.FaultTolerance.<>c__DisplayClass1.<retry>b__0()
   at chocolatey.infrastructure.tolerance.FaultTolerance.retry[T](Int32 numberOfTries, Func`1 function, Int32 waitDurationMilliseconds, Int32 increaseRetryByMilliseconds)
This is try 2/3. Retrying after 400 milliseconds.
 Error converted to warning:
 System.UnauthorizedAccessException: Access to the path 'RBHook.dll' is denied.
   at System.IO.Directory.DeleteHelper(String fullPath, String userPath, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.Directory.Delete(String fullPath, String userPath, Boolean recursive, Boolean checkHost)
   at chocolatey.infrastructure.tolerance.FaultTolerance.<>c__DisplayClass1.<retry>b__0()
   at chocolatey.infrastructure.tolerance.FaultTolerance.retry[T](Int32 numberOfTries, Func`1 function, Int32 waitDurationMilliseconds, Int32 increaseRetryByMilliseconds)
Maximum tries of 3 reached. Throwing error.
Attempted to remove 'C:\ProgramData\chocolatey\lib-bkp\rbtray' but had an error::
 System.UnauthorizedAccessException: Access to the path 'RBHook.dll' is denied.
   at System.IO.Directory.DeleteHelper(String fullPath, String userPath, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.Directory.Delete(String fullPath, String userPath, Boolean recursive, Boolean checkHost)
   at chocolatey.infrastructure.tolerance.FaultTolerance.<>c__DisplayClass1.<retry>b__0()
   at chocolatey.infrastructure.tolerance.FaultTolerance.retry[T](Int32 numberOfTries, Func`1 function, Int32 waitDurationMilliseconds, Int32 increaseRetryByMilliseconds)
   at chocolatey.infrastructure.tolerance.FaultTolerance.<>c__DisplayClass4.<try_catch_with_logging_exception>b__3()
   at chocolatey.infrastructure.tolerance.FaultTolerance.try_catch_with_logging_exception[T](Func`1 function, String errorMessage, Boolean throwError, Boolean logWarningInsteadOfError)
```

**Comment by ferventcoder** [comment]
Is their a service or something locking the file?

**Comment by ferventcoder** [comment]
Can you include the full log please?

**Comment by ferventcoder** [comment]
We tend to move the files over to lib-bkp, which allows locked files to go because they are not being deleted. Then when it tries to delete the file, something is still holding a lock on it. Perhaps it would be better to warn in that folder...

**Comment by vorou** [comment]
> Can you include the full log please?
https://gist.github.com/vorou/4c084bb3165c26412a08

**Comment by vorou** [comment]
Found that you need admin rights to delete the DLL. Didn't notice it earlier since I was in admin PowerShell. When I'm trying to delete the file in Explorer, it shows the dialog asking for permissions, and then remove the file.

So, I created `removedll.ps1`:

```
Write-Output "Removing the DLLs..."
rm 'C:\ProgramData\chocolatey\lib\rbtray\tools\64bit\RBHook.dll'
rm 'C:\ProgramData\chocolatey\lib-bkp\rbtray\tools\64bit\RBHook.dll'
Write-Output "DLLs removed."
```

And I call it from uninstall script like so:

```
$toolsDir ="$(Split-Path -parent $MyInvocation.MyCommand.Definition)"
Start-ChocolateyProcessAsAdmin "& $($toolsDir)\removedll.ps1"
```

It finishes w/o any errors, but then same stuff happens. Choco can't remove the file in lib-bkp, and afterwards I can see the files just like nobody was removing them. It's sooo strange.

Here's [the full log](https://gist.github.com/3c92a536a9acb0fe6c79) for this one.

**Comment by vorou** [comment]
I almost gave up on getting rid of the error messages, but noticed that the uninstall script doesn't remove `lib\rbtray` folder. So, after uninstall I have `lib\rbtray` folder untouched, and a newly created copy of `RBHook.dll` in `lib-bkp\rbtray`.

Here's my [`chocolateyUninstall.ps1`](https://gist.github.com/1d6d58c05ff7d1ebe0bc). What am I doing wrong?

**Comment by qzo** [comment]
I have the exact same problem with a ZIP app containing a read-only file (. I get warning during install and uninstall processes, but after the installation is finished, everything seems to be OK. But after uninstall, the read-only file is still present in both lib and lib-bkp.
The package in question is:
https://github.com/qzo/chocolatey-packages/tree/6b892168b853ce537cf9f1b25aa2009f5f266a5a/packages/manual/farr_m/find-and-run-robot.portable
Logs:
https://gist.github.com/qzo/e4965ecdbb6e8ca773a3
https://gist.github.com/qzo/5e77b1844f87b4bef915

**Comment by ferventcoder** [comment]
@qzo The stop process - how does that work with upgrades?

**Comment by ferventcoder** [comment]
Is it possible that the process isn't fully stopped yet, so it's still locking the files. You may want to insert about a second or two of sleep in there to ensure that the process is fully stopped.

**Comment by ferventcoder** [comment]
And do a try catch instead of silently continuing, so you can log out any errors that are found, like the process not shutting down.

**Comment by qzo** [comment]
But the application process isn't running at all - it can't run before the installation and I also checked it didn't run before uninstall, but I got the warnings anyway. I'm 99 % sure it has to do with the file in question having read-only attribute.

The Stop-Process isn't really relevant - I just didn't want upgrade / uninstall to fail due to the application still running. But I just found out uninstall causes the process to close, so maybe it's not necessary after all.

**Comment by ferventcoder** [comment]
Good point on read only attributes. Let's file a new issue on that.

**Comment by qzo** [comment]
Here you go: #338

**Comment by ferventcoder** [comment]
Looks like this was switched to a warning awhile ago - https://github.com/chocolatey/choco/commit/217e8fa589319b00e2a3b7bb67b383619313ef4d

**Comment by ferventcoder** [comment]
I'm going to close this one as part of #338 fixes. If this comes back up as not fixed in 0.9.9.9, please feel free to comment and we can reopen.
